---
title: "Analyze GTex Data with Brian's Summary Data"
author: "Chris McKennan"
date: '2016-01-25'
output: html_document
---

**Last updated:** `r Sys.Date()`

**Code version:** `r system("git log -1 --format='%H'", intern = TRUE)`

```{r chunk-options, include=FALSE}
source("chunk-options.R")
```

## Objectives

This file takes Brian's muscle skeletal (correlation threshold = 0.80) GTex summary files and performs a Gibbs sampler to estimate the posterior probability each gene in the network is UNAFFECTED by the genotype at the SNP. The results are store in '../output', with the same directory structure as the summary data. The only input parameter is the input directory name with summary data and the output directory, which stores the Gibbs sampler posterior probability estimates. After the code finishes, a Perl script will modify the Gibbs sampler output so they have the same information as the summary data files.


## Grab functions

Source Gibbs Sampler and Bayes Factor functions.

```{r Grab R functions}
source("../R/SimulateNetworkExpression.R")  #Compute Bayes factors
source("../R/directionality_CGM.R")    #Gibbs Sampler
```

## The directories to grab and store output files

Directory in which to perform the analysis. Note that this is the muscle skeletal. Results are saved in output, with the same directory structure as the summary data were stored in data

```{r Initialize Directories}
directory.in <- '../data/cis_summary_data/muscleskeletal/thresh_80'
directory.out <- '../output/cis_summary_data/muscleskeletal/thresh_80'
run_gibbs = FALSE;   #Whether or not to run the Gibbs sampler. If not, I just look for the file in 'directory.out' to get the necessary information 
```

## Create directory structure to house output files

```{r Create Directory Structure}
if (!file.exists(directory.out)) {
  dir.string <- strsplit(directory.out, split="/")[[1]]
  tmp.string = dir.string[1]
  for (i in 2:length(dir.string)) {
    tmp.string = file.path(tmp.string, dir.string[i])
    if (!file.exists(tmp.string)) {
      dir.create(tmp.string)
    }
  }
}
```


## Run the analysis on all of the files in the directory

```{r GibbsSampler}
all_files <- list.files(directory.in)
N.genes <- rep(0, length(all_files))    #Number of genes in each network
PUP.source <- rep(0, length(all_files))  #Posterior probability the gene cis to the eQTL is unaffected
all.post.probs <- c()   #To be used to generate a ROC plot

count = 1
for (file in all_files) {
  file.path <- paste(directory.in, file, sep="/")
  all_data = readLines(file.path)
  tissue <- strsplit(all_data[1], split='\t', perl=T)[[1]][2]
	n.ind <- as.numeric(strsplit(all_data[2], split='\t', perl=T)[[1]][2])   #Number of independent measurements
	chr <- as.numeric(strsplit(all_data[3], split='\t', perl=T)[[1]][2])    #Chromosome number
	gene <- strsplit(all_data[4], split='\t', perl=T)[[1]][2]    #Gene of interest
	SNP <- strsplit(all_data[5], split='\t', perl=T)[[1]][2]    #eQTL of interest
	Gene.names <- strsplit(all_data[6], split='\t', perl=T)[[1]][2:length(strsplit(all_data[6], split='\t', perl=T)[[1]])]   #Column names for Y'Y
	n.genes <- length(Gene.names)    #Number of genes in the network
  N.genes[count] = n.genes
	
	YtY <- array(NA, dim=c(n.genes, n.genes))   #Y'Y
	for (r in 1:n.genes) {
	  YtY[r,] <- as.numeric(strsplit(all_data[7+r], split='\t', perl=T)[[1]])
	}
	
	sxx <- as.numeric(strsplit(all_data[8+n.genes], split='\t', perl=T)[[1]][2])   #X'X, a scalar
	YtX <- as.numeric(strsplit(all_data[10+n.genes], split='\t', perl=T)[[1]])    #Y'X, a vector
	
	suff.stat <- list(SYY = YtY/n.ind, sxx = sxx/n.ind, SYX = YtX/n.ind, SY1 = rep(0, n.genes), mu.g = 0)
	
	if (n.genes <= 15) {
	  n.iter <- 3000
	  n.burn <- 1000
	} else {
	  if (n.genes <= 25) {
	    n.iter <- 4000
	    n.burn <- 2000
	  } else {
	    n.iter <- 5000
	    n.burn <- 2500
	  }
	}
	sigma.a <- c(0.1, 0.4)
	weights.sigma <- c(1, 1)
	
  out.file = file.path(directory.out, sub("(summary)", '\\1.gibbs', file, perl=T))
  if (run_gibbs) {
    gibbs <- Gibbs.dir(n.iter, n.burn, suff.stat, which(Gene.names == gene), n.ind, sigma.a, weights.sigma, n.genes-1)
    PUP.source[count] = gibbs$post.mean[gibbs$D.ind]
  
    gibbs.mat <- rbind(gibbs$post.mean)
    colnames(gibbs.mat) <- Gene.names
    write.table(gibbs.mat, out.file, col.names=T, row.names=F, append=F, quote=F, sep="\t")
    all.post.probs <- c(all.post.probs, gibbs$post.mean[-gibbs$D.ind])
  } else {
    gibbs.file = readLines(out.file)
    tmp.probs <- as.numeric(strsplit(gibbs.file[8], split='\t', perl=T)[[1]])
    PUP.source[count] = tmp.probs[which(Gene.names == gene)]
    all.post.probs <- c(all.post.probs, tmp.probs[-which(Gene.names == gene)])
  }
  
  #print(paste0(as.character(count), " of ", as.character(length(all_files))))
  count = count + 1
}
```

Plot the posterior probability a gene is affected by SNP as a function of the network size:
```{r PlotPAP}
plot(N.genes, 1-PUP.source, main='Posterior Probability of Source Gene as a Function of Network Size', xlab='Number of Genes in Network', ylab='Posterior Probability Source Gene is Affected by SNP')

jpeg('AnalyzeSummaryData_muscleskeletal80_NetworkSize.jpeg', width = 550)
plot(N.genes, 1-PUP.source, main='Posterior Probability of Source Gene as a Function of Network Size', xlab='Number of Genes in Network', ylab='Posterior Probability Source Gene is Affected by SNP')
dev.off()
```

Plot an estimated ROC curve:
```{r ROC}
all.post.probs <- sort(all.post.probs)   #Posterior prob a gene is NOT a trans eQTL
est.trans <- sum(1-all.post.probs)   #Expected number of trans genes affected (given the data)

est.FDR <- rep(0, length(all.post.probs))
est.FDR[1] <- all.post.probs[1]
est.sens <- rep(0, length(all.post.probs))
est.sens[1] <- (1 - all.post.probs[1])/est.trans
for (i in 2:length(est.FDR)) {
  est.FDR[i] <- ((i-1)*est.FDR[i-1] + all.post.probs[i])/i
  est.sens[i] <- est.sens[i-1] + (1 - all.post.probs[i])/est.trans
}
plot(est.FDR, est.sens, type="l", main='Estimated ROC Curve for Affected Genes Trans to SNP in Muscle Skeletal', xlab="Estimated (conditional) False Discovery Rate", ylab='Estimated Sensitivity', xlim=c(0, max(0.6, est.FDR)))
lines(seq(max(est.FDR), 1, length=100), rep(1, 100))

jpeg('AnalyzeSummaryData_muscleskeletal80_ROC.jpeg', width = 550)
plot(est.FDR, est.sens, type="l", main='Estimated ROC Curve for Affected Genes Trans to SNP in Muscle Skeletal', xlab="Estimated (conditional) False Discovery Rate", ylab='Estimated Sensitivity', xlim=c(0, max(0.6, est.FDR)))
lines(seq(max(est.FDR), 1, length=100), rep(1, 100))
dev.off()
```

Histgram of posterior probabilities that a gene trans to SNP is affected by SNP
```{r Histogram}
hist(1-all.post.probs, breaks=30, xlab="Posterior Probability Gene Trans to SNP is AFFECTED by SNP", ylab="Frequency", main="Histogram of Posterior Probabilities a Trans Gene is AFFECTED by SNP", xlim=c(0,1))

jpeg('AnalyzeSummaryData_muscleskeletal80_Hist.jpeg', width = 550)
hist(1-all.post.probs, breaks=30, xlab="Posterior Probability Gene Trans to SNP is AFFECTED by SNP", ylab="Frequency", main="Histogram of Posterior Probabilities a Trans Gene is AFFECTED by SNP", xlim=c(0,1))
dev.off()
```


## Add additional information to output files
If the information already exists, OrganizeGibbs.pl leaves the file alone

```bash
perl OrganizeGibbs.pl ../data/cis_summary_data/muscleskeletal/thresh_80 ../output/cis_summary_data/muscleskeletal/thresh_80
```

## Session information

```{r info}
sessionInfo()
```
