---
title: "Analyze Consistency of trans-eQTLs Across Tissues"
author: "Chris McKennan"
date: '2016-03-08'
output: html_document
---

**Last updated:** `r Sys.Date()`

**Code version:** `r system("git log -1 --format='%H'", intern = TRUE)`

```{r chunk-options, include=FALSE}
source("chunk-options.R")
```

```{r Packages, include=FALSE}
library('knitr')
library('printr')
```

## Objectives

This file analyzes the results of the Gibbs sampler run on Brian's 0.8 thresholded networks. Specifically, we look for (in)consistency of trans-eQTLs across tissues. We also identify a few networks that are known, and validate our resutls on those networks.

## Analyze Data

This perl script will look at all of the output files in the ../output/"tissue"/thresh80 look at the overlap of results in tissues. The sctipt also prints out all of the gene A -> gene B connections that are shared across all tissues for confirmatory analysis.

## Trans-eQTLs present in all 3 tissues with PEP $\leq$ 0.05:
```{r Perl05}
system("perl ConsistencyTissues.pl 0.05", intern=T)
```


## Trans-eQTLs present in all 3 tissues with PEP $\leq$ 0.1:
```{r Perl1}
system("perl ConsistencyTissues.pl 0.1", intern=T)
```

## Trans-eQTLs present in all 3 tissues with PEP $\leq$ 0.2:
```{r Perl2}
system("perl ConsistencyTissues.pl 0.2", intern=T)
```

## Some Verification
ENSG00000173372, ENSG00000173369 and ENSG00000159189 are C1Q, chains A, B and C (complement component 1, q subcomponent). Note however that these are located on adjacent regions of chromosome 1 and may not be that interesting.

We identify a gene network of 3 histone proteins (thresh 0.60) with HLA-DRB1 (ENSG00000196126) as the anchor and HLA-DQB1, HLA-DRB5 as downstream targets. Again, these are adjecent to one another on chr 6 and may not be very interesting.

We also identify ENSG00000161970 (RPL26), which seems to affect the expression of ENSG00000147604 (RPL7), ENSG00000138326 (RPS24), ENSG00000109475 (RPL34), ENSG00000145425 (RPS3A). These genes are located on different chromosomes and all of these genes are protein coding and correspond to ribosomal proteins. It is reasuring that these genes may have similar function.

Zhang et al (BMC Genomics, 2014) found strong evidence that RPL14 (Chr 1) is involved in a large regulatory network. This is interesting because it is another ribosomal protein network. Brian identifies this as a source gene in whole blood (ENSG00000188846) with a correlation threshold of 0.60. However, we do not see evidence that it influences the expression of downstream genes. It is worth noting that they used a different SNP as the cis-eQTL, but this shouldn't make a difference if Brian's SNP is in high LD with their SNP.

## Global Tissue Sharing Statistics

Below is a table giving the number of Trans-eQTLs that appear in ONLY 1 tissue, 2 tissues or in all 3 tissues. I defined a trans-eQTL if it had a posterior probability of being unaffected less than or equal to 0.1.

```{r TissuePlots}
path.data <- 'TissueCount_pep0.1.txt'
x <- read.table(path.data, sep="\t", dec=".", check.names=F, header=F, row.names=1)
colnames(x) <- c('Number of Trans-eQTL Targets')

knitr::kable(x, caption="Trans-eQTL Sharing Across Tissues", longtable=F)
```

Note how trans-eQTLs are not frequently shared across tissues. This is consistent with previous findings, although it makes results difficult to verify.

## Session information

```{r info}
sessionInfo()
```
